zk/groth16/src/verifier.rs

```diff
@@ -14,6 +21,60 @@ pub fn groth16_verify<E: Pairing>(
     Groth16::<E, LibsnarkReduction>::verify_proof(vk.as_ref(), &proof, public_inputs)
 }
 
+pub fn groth16_batch_verify(
+    vk: &PreparedVerificationKey<Bn254>,
+    proofs: &[Proof<Bn254>],
+    public_inputs: &[Vec<Fr>],
+) -> R1CSResult<bool> {
+    let mut rng = thread_rng();
+    let r = Fr::rand(&mut rng);
+    let r_roots = compute_r_powers(r, proofs.len());
+    let r_sum: Fr = r_roots.iter().sum();
+
+    let pis_c: Vec<G1Affine> = proofs.iter().map(|proof| proof.pi_c).collect();
+
+    let batched_pi_c = G1Projective::msm(&pis_c, &r_roots).unwrap().into_affine();
+
+    let batched_public_inputs: Vec<Fr> = std::iter::once(r_sum)
+        .chain((0..public_inputs[0].len()).map(|i| {
+            r_roots
+                .iter()
+                .zip(public_inputs.iter())
+                .map(|(r, pi)| *r * pi[i])
+                .sum()
+        }))
+        .collect();
+
+    let batched_ic = G1Projective::msm(&vk.vk.vk.gamma_abc_g1, &batched_public_inputs)
+        .unwrap()
+        .into_affine();
+
+    let mut g1_terms: Vec<_> = Vec::with_capacity(proofs.len() + 3);
+    let mut g2_terms: Vec<_> = Vec::with_capacity(proofs.len() + 3);
+
+    for (i, proof) in proofs.iter().enumerate() {
+        g1_terms.push(proof.pi_a.mul(r_roots[i]).into_affine());
+        g2_terms.push(proof.pi_b.into());
+    }
+    g1_terms.push(vk.vk.vk.alpha_g1.mul(r_sum).neg().into());
+    g2_terms.push(vk.vk.vk.beta_g2.into());
+    g1_terms.push(batched_ic);
+    g2_terms.push(vk.vk.gamma_g2_neg_pc.clone());
+    g1_terms.push(batched_pi_c);
+    g2_terms.push(vk.vk.delta_g2_neg_pc.clone());
+
+    let qap = Bn254::multi_miller_loop(g1_terms, g2_terms);
+
+    let test = Bn254::final_exponentiation(qap).ok_or(SynthesisError::UnexpectedIdentity)?;
```

Comment:
Can we use the pairing API here?
```suggestion
    let test = Bn254::multi_pairing(g1_terms, g2_terms);
```